server/auth/auth.go

```diff
@@ -657,44 +657,73 @@ func (auth *AuthManager) UserSettlingLimit(user account.AccountID, mkt *dex.Mark
 	return limit
 }
 
-// userScore computes the user score from the user's recent match outcomes and
-// preimage history. This must be called with the violationMtx locked.
-func (auth *AuthManager) userScore(user account.AccountID) (score int32) {
-	if outcomes, found := auth.matchOutcomes[user]; found {
-		for v, count := range outcomes.binViolations() {
+func integrateOutcomes(matchOutcomes *latestMatchOutcomes, preimgOutcomes *latestPreimageOutcomes) (score, successCount, piMissCount int32) {
+	if matchOutcomes != nil {
+		matchCounts := matchOutcomes.binViolations()
+		for v, count := range matchCounts {
 			score += v.Score() * int32(count)
 		}
+		successCount = int32(matchCounts[ViolationSwapSuccess])
 	}
-	if outcomes, found := auth.preimgOutcomes[user]; found {
-		score += ViolationPreimageMiss.Score() * outcomes.misses()
+	if preimgOutcomes != nil {
+		piMissCount = preimgOutcomes.misses()
+		score += ViolationPreimageMiss.Score() * piMissCount
 	}
 	return
 }
 
-func (auth *AuthManager) registerMatchOutcome(user account.AccountID, misstep NoActionStep, mmid db.MarketMatchID, value uint64, refTime time.Time) int32 {
+// userScore computes an authenticated user's score from their recent match
+// outcomes and preimage history. They must have entries in the outcome maps.
+// Use loadUserScore to compute score from history in DB. This must be called
+// with the violationMtx locked.
+func (auth *AuthManager) userScore(user account.AccountID) (score int32) {
+	score, _, _ = integrateOutcomes(auth.matchOutcomes[user], auth.preimgOutcomes[user])
+	return score
+}
+
+func (auth *AuthManager) registerMatchOutcome(user account.AccountID, misstep NoActionStep, mmid db.MarketMatchID, value uint64, refTime time.Time) (score int32) {
 	violation := misstep.Violation()
 
 	auth.violationMtx.Lock()
 	defer auth.violationMtx.Unlock()
-	outcomes, found := auth.matchOutcomes[user]
-	if !found {
-		outcomes = newLatestMatchOutcomes(scoringMatchLimit)
-		auth.matchOutcomes[user] = outcomes
-	}
```

Comment:
This is the gist of the bug.  I suspect it was written with the thinking "no outcomes, so this must be a new user" but the absence of an entry really means offline user.